README.md
![GitHub](https://img.shields.io/github/license/zevisert/fastatus)
![npm bundle size](https://img.shields.io/bundlephobia/minzip/fastatus)
![npm](https://img.shields.io/npm/v/fastatus)

## What is `fastatus`?

`fastatus` (_FastAPI Status_) is an isomorphic helper library that provides an autocomplete friendly list of HTTP and Websocket status codes. It is modeled after the [`starlette.status`](https://github.com/encode/starlette/blob/25a52fe52a1fa4f55da4096fb0c7e6f6de0ae980/starlette/status.py) module from python used by [FastAPI](https://fastapi.tiangolo.com/).

My reason to release yet another http status code mapping package was for the specific names for the enumerated members. I wanted a package with both the number and the code name in the properties. This is the same _raison d'être_ for `starlette.status`, and it really helps when it comes to editor support:

![fastatus autocomplete in action](./demo.png)

## Installation

Install the [`fastatus` package](https://www.npmjs.com/package/fastatus) from npm using your favorite package manager.

```console
$ pnpm add fastatus
$ yarn add fastatus
$ npm install fastatus
```

## Usage

HTTP status codes are provided as an enum.

```typescript
import status from "fastatus";

console.log(status.HTTP_426_UPGRADE_REQUIRED);
// => 426

// Reverse mappings are available
console.log(status[508]);
// => 'HTTP_508_LOOP_DETECTED'
```

In addition to what's found in `starlette.status`, you can also get the IANA text description.

```typescript
import { description } from "fastatus";
console.log(description.HTTP_505_HTTP_VERSION_NOT_SUPPORTED);
// => "HTTP Version Not Supported"
```

## Status Codes

All assigned [http status codes](https://www.iana.org/assignments/http-status-codes/http-status-codes.xhtml#http-status-codes-1) and [websocket close codes](https://www.iana.org/assignments/websocket/websocket.xml#close-code-number) are included.
| Enum | Value | Description |
| ------------------------------------------ | --------- | ----------------------------------------------------------------------------------------------------- |
| `HTTP_100_CONTINUE` | 100 | Continue |
| `HTTP_101_SWITCHING_PROTOCOLS` | 101 | Switching Protocols |
| `HTTP_102_PROCESSING` | 102 | Processing |
| `HTTP_103_EARLY_HINTS` | 103 | Early Hints |
| | 104-199 | Unassigned |
| `HTTP_200_OK` | 200 | OK |
| `HTTP_201_CREATED` | 201 | Created |
| `HTTP_202_ACCEPTED` | 202 | Accepted |
| `HTTP_203_NON_AUTHORITATIVE_INFORMATION` | 203 | Non-Authoritative Information |
| `HTTP_204_NO_CONTENT` | 204 | No Content |
| `HTTP_205_RESET_CONTENT` | 205 | Reset Content |
| `HTTP_206_PARTIAL_CONTENT` | 206 | Partial Content |
| `HTTP_207_MULTI_STATUS` | 207 | Multi-Status |
| `HTTP_208_ALREADY_REPORTED` | 208 | Already Reported |
| | 209-225 | Unassigned |
| `HTTP_226_IM_USED` | 226 | IM Used |
| | 227-299 | Unassigned |
| `HTTP_300_MULTIPLE_CHOICES` | 300 | Multiple Choices |
| `HTTP_301_MOVED_PERMANENTLY` | 301 | Moved Permanently |
| `HTTP_302_FOUND` | 302 | Found |
| `HTTP_303_SEE_OTHER` | 303 | See Other |
| `HTTP_304_NOT_MODIFIED` | 304 | Not Modified |
| `HTTP_305_USE_PROXY` | 305 | Use Proxy |
| `HTTP_306_RESERVED` | 306 | (Unused) |
| `HTTP_307_TEMPORARY_REDIRECT` | 307 | Temporary Redirect |
| `HTTP_308_PERMANENT_REDIRECT` | 308 | Permanent Redirect |
| | 309-399 | Unassigned |
| `HTTP_400_BAD_REQUEST` | 400 | Bad Request |
| `HTTP_401_UNAUTHORIZED` | 401 | Unauthorized |
| `HTTP_402_PAYMENT_REQUIRED` | 402 | Payment Required |
| `HTTP_403_FORBIDDEN` | 403 | Forbidden |
| `HTTP_404_NOT_FOUND` | 404 | Not Found |
| `HTTP_405_METHOD_NOT_ALLOWED` | 405 | Method Not Allowed |
| `HTTP_406_NOT_ACCEPTABLE` | 406 | Not Acceptable |
| `HTTP_407_PROXY_AUTHENTICATION_REQUIRED` | 407 | Proxy Authentication Required |
| `HTTP_408_REQUEST_TIMEOUT` | 408 | Request Timeout |
| `HTTP_409_CONFLICT` | 409 | Conflict |
| `HTTP_410_GONE` | 410 | Gone |
| `HTTP_411_LENGTH_REQUIRED` | 411 | Length Required |
| `HTTP_412_PRECONDITION_FAILED` | 412 | Precondition Failed |
| `HTTP_413_REQUEST_ENTITY_TOO_LARGE` | 413 | Content Too Large |
| `HTTP_414_REQUEST_URI_TOO_LONG` | 414 | URI Too Long |
| `HTTP_415_UNSUPPORTED_MEDIA_TYPE` | 415 | Unsupported Media Type |
| `HTTP_416_REQUESTED_RANGE_NOT_SATISFIABLE` | 416 | Range Not Satisfiable |
| `HTTP_417_EXPECTATION_FAILED` | 417 | Expectation Failed |
| `HTTP_418_IM_A_TEAPOT` | 418 | (Unused) |
| | 419-420 | Unassigned |
| `HTTP_421_MISDIRECTED_REQUEST` | 421 | Misdirected Request |
| `HTTP_422_UNPROCESSABLE_ENTITY` | 422 | Unprocessable Content |
| `HTTP_423_LOCKED` | 423 | Locked |
| `HTTP_424_FAILED_DEPENDENCY` | 424 | Failed Dependency |
| `HTTP_425_TOO_EARLY` | 425 | Too Early |
| `HTTP_426_UPGRADE_REQUIRED` | 426 | Upgrade Required |
| | 427 | Unassigned |
| `HTTP_428_PRECONDITION_REQUIRED` | 428 | Precondition Required |
| `HTTP_429_TOO_MANY_REQUESTS` | 429 | Too Many Requests |
| | 430 | Unassigned |
| `HTTP_431_REQUEST_HEADER_FIELDS_TOO_LARGE` | 431 | Request Header Fields Too Large |
| | 432-450 | Unassigned |
| `HTTP_451_UNAVAILABLE_FOR_LEGAL_REASONS` | 451 | Unavailable For Legal Reasons |
| | 452-499 | Unassigned |
| `HTTP_500_INTERNAL_SERVER_ERROR` | 500 | Internal Server Error |
| `HTTP_501_NOT_IMPLEMENTED` | 501 | Not Implemented |
| `HTTP_502_BAD_GATEWAY` | 502 | Bad Gateway |
| `HTTP_503_SERVICE_UNAVAILABLE` | 503 | Service Unavailable |
| `HTTP_504_GATEWAY_TIMEOUT` | 504 | Gateway Timeout |
| `HTTP_505_HTTP_VERSION_NOT_SUPPORTED` | 505 | HTTP Version Not Supported |
| `HTTP_506_VARIANT_ALSO_NEGOTIATES` | 506 | Variant Also Negotiates |
| `HTTP_507_INSUFFICIENT_STORAGE` | 507 | Insufficient Storage |
| `HTTP_508_LOOP_DETECTED` | 508 | Loop Detected |
| | 509 | Unassigned |
| `HTTP_510_NOT_EXTENDED` | 510 | Not Extended (OBSOLETED) |
| `HTTP_511_NETWORK_AUTHENTICATION_REQUIRED` | 511 | Network Authentication Required |
| | 512-599 | Unassigned |
| `WS_1000_NORMAL_CLOSURE` | 1000 | Normal Closure |
| `WS_1001_GOING_AWAY` | 1001 | Going Away |
| `WS_1002_PROTOCOL_ERROR` | 1002 | Protocol error |
| `WS_1003_UNSUPPORTED_DATA` | 1003 | Unsupported Data |
| | 1004 | Reserved |
| `WS_1005_NO_STATUS_RCVD` | 1005 | No Status Rcvd |
| `WS_1006_ABNORMAL_CLOSURE` | 1006 | Abnormal Closure |
| `WS_1007_INVALID_FRAME_PAYLOAD_DATA` | 1007 | Invalid frame payload data |
| `WS_1008_POLICY_VIOLATION` | 1008 | Policy Violation |
| `WS_1009_MESSAGE_TOO_BIG` | 1009 | Message Too Big |
| `WS_1010_MANDATORY_EXT` | 1010 | Mandatory Ext. |
| `WS_1011_INTERNAL_ERROR` | 1011 | Internal Error |
| `WS_1012_SERVICE_RESTART` | 1012 | Service Restart |
| `WS_1013_TRY_AGAIN_LATER` | 1013 | Try Again Later |
| `WS_1014_BAD_GATEWAY` | 1014 | The server was acting as a gateway or proxy and received an invalid response from the upstream server |
| `WS_1015_TLS_HANDSHAKE` | 1015 | TLS handshake |
| | 1016-2999 | Unassigned |
| | 3000 | Unauthorized |
| | 3001-3999 | Unassigned |
| | 4000-4999 | Reserved for Private Use |

## Acknowledgements

- Forked from [rmuchall/http-status-ts](https://github.com/rmuchall/http-status-ts)
- [encode/starlette](https://github.com/encode/starlette/) for the great naming convention


CHANGELOG.md
# Changelog

#### v1.0.0

- chore: replace project configuration
- feat: replace src with from starlette and iana data
- fork: github.com/rmuchall/http-status-ts for project structure


test/index.test.ts
import { status } from "../src";
import { description } from "../src";
import _default from "../src";

test("can lookup status", () => {
  expect(status.HTTP_200_OK).toBe(200);
  expect(status.HTTP_418_IM_A_TEAPOT).toBe(418);
  expect(status["HTTP_302_FOUND"]).toBe(302);
  expect(status[303]).toBe("HTTP_303_SEE_OTHER");
});

test("can lookup descriptions", () => {
  expect(description.HTTP_101_SWITCHING_PROTOCOLS).toBe("Switching Protocols");
  expect(description.HTTP_206_PARTIAL_CONTENT).toBe("Partial Content");
  expect(description["HTTP_304_NOT_MODIFIED"]).toBe("Not Modified");
});

function keys(e: Record<string, unknown>, prefix = ""): string[] {
  return Object.keys(e).filter((key) => key.startsWith(prefix));
}

test("is fully described", () => {
  // 63 HTTP codes
  expect(keys(status, "HTTP").length).toBe(63);

  // 15 websocket codes
  expect(keys(status, "WS").length).toBe(15);

  // Times 2 because of TS enum implementation of reverse mapping for non-string enums
  expect(keys(status).length).toBe((63 + 15) * 2);

  // Same number of descriptions as codes
  expect(keys(status).length / 2).toBe(keys(description).length);
});


test/http-status.test.ts
import {HttpStatus, httpStatusTextByCode} from "../src/HttpStatus";

test("textByCode()", () => {
    expect(HttpStatus.OK).toBe(200);
    expect(httpStatusTextByCode(200)).toBe("OK");
    expect(httpStatusTextByCode("200")).toBe("OK");
    expect(() => httpStatusTextByCode("xxxx")).toThrow();
    expect(() => httpStatusTextByCode(9999)).toThrow();
    expect(() => httpStatusTextByCode(undefined as any)).toThrow();
    expect(() => httpStatusTextByCode(null as any)).toThrow();
    expect(() => httpStatusTextByCode({} as any)).toThrow();
});


.eslintrc.js
module.exports = {
  root: true,
  env: {
    node: true,
  },
  extends: [
    "eslint:recommended",
    "plugin:@typescript-eslint/recommended",
    "plugin:@typescript-eslint/recommended-requiring-type-checking",
  ],
  parser: "@typescript-eslint/parser",
  parserOptions: {
    tsconfigRootDir: __dirname,
    project: "tsconfig.json",
  },
  plugins: ["@typescript-eslint"],
  ignorePatterns: [".eslintrc.js", "jest.config.js"],
  rules: {
    // Style - Enable
    quotes: ["error", "double", { allowTemplateLiterals: true }],
    semi: ["error", "always"],
    // Nodejs - Enable
    "require-await": "error",
    "no-return-await": "error",
    "@typescript-eslint/explicit-module-boundary-types": "error",
    "@typescript-eslint/unbound-method": ["error", { ignoreStatic: true }],
    // NodeJs - Disable
    "@typescript-eslint/no-unused-vars": "off",
    "@typescript-eslint/no-inferrable-types": "off",
    "@typescript-eslint/no-unsafe-member-access": "off",
    "@typescript-eslint/no-unsafe-assignment": "off",
    "@typescript-eslint/no-unsafe-call": "off",
    "@typescript-eslint/no-unsafe-argument": "off",
    "@typescript-eslint/restrict-template-expressions": "off",
    "@typescript-eslint/no-non-null-assertion": "off",
    "@typescript-eslint/no-explicit-any": "off",
  },
};


src/HttpStatus.ts
// https://www.iana.org/assignments/http-status-codes/http-status-codes.xhtml
export enum HttpStatus {
    CONTINUE = 100,
    SWITCHING_PROTOCOLS = 101,
    PROCESSING = 102,
    EARLY_HINTS = 103,
    OK = 200,
    CREATED = 201,
    ACCEPTED = 202,
    NON_AUTHORITATIVE_INFORMATION = 203,
    NO_CONTENT = 204,
    RESET_CONTENT = 205,
    PARTIAL_CONTENT = 206,
    MULTI_STATUS = 207,
    ALREADY_REPORTED = 208,
    IM_USED = 226,
    MULTIPLE_CHOICES = 300,
    MOVED_PERMANENTLY = 301,
    FOUND = 302,
    SEE_OTHER = 303,
    NOT_MODIFIED = 304,
    USE_PROXY = 305,
    TEMPORARY_REDIRECT = 307,
    PERMANENT_REDIRECT = 308,
    BAD_REQUEST = 400,
    UNAUTHORIZED = 401,
    PAYMENT_REQUIRED = 402,
    FORBIDDEN = 403,
    NOT_FOUND = 404,
    METHOD_NOT_ALLOWED = 405,
    NOT_ACCEPTABLE = 406,
    PROXY_AUTHENTICATION_REQUIRED = 407,
    REQUEST_TIMEOUT = 408,
    CONFLICT = 409,
    GONE = 410,
    LENGTH_REQUIRED = 411,
    PRECONDITION_FAILED = 412,
    PAYLOAD_TOO_LARGE = 413,
    URI_TOO_LONG = 414,
    UNSUPPORTED_MEDIA_TYPE = 415,
    RANGE_NOT_SATISFIABLE = 416,
    EXPECTATION_FAILED = 417,
    MISDIRECTED_REQUEST = 421,
    UNPROCESSABLE_ENTITY = 422,
    LOCKED = 423,
    FAILED_DEPENDENCY = 424,
    TOO_EARLY = 425,
    UPGRADE_REQUIRED = 426,
    PRECONDITION_REQUIRED = 428,
    TOO_MANY_REQUESTS = 429,
    REQUEST_HEADER_FIELDS_TOO_LARGE = 431,
    UNAVAILABLE_FOR_LEGAL_REASONS = 451,
    INTERNAL_SERVER_ERROR = 500,
    NOT_IMPLEMENTED = 501,
    BAD_GATEWAY = 502,
    SERVICE_UNAVAILABLE = 503,
    GATEWAY_TIMEOUT = 504,
    HTTP_VERSION_NOT_SUPPORTED = 505,
    VARIANT_ALSO_NEGOTIATES = 506,
    INSUFFICIENT_STORAGE = 507,
    LOOP_DETECTED = 508,
    NOT_EXTENDED = 510,
    NETWORK_AUTHENTICATION_REQUIRED = 511
}

const httpStatusText: Record<string, string> = {
    [HttpStatus.CONTINUE]: "Continue",
    [HttpStatus.SWITCHING_PROTOCOLS]: "Switching Protocols",
    [HttpStatus.PROCESSING]: "Processing",
    [HttpStatus.EARLY_HINTS]: "Early Hints",
    [HttpStatus.OK]: "OK",
    [HttpStatus.CREATED]: "Created",
    [HttpStatus.ACCEPTED]: "Accepted",
    [HttpStatus.NON_AUTHORITATIVE_INFORMATION]: "Non-Authoritative Information",
    [HttpStatus.NO_CONTENT]: "No Content",
    [HttpStatus.RESET_CONTENT]: "Reset Content",
    [HttpStatus.PARTIAL_CONTENT]: "Partial Content",
    [HttpStatus.MULTI_STATUS]: "Multi-Status",
    [HttpStatus.ALREADY_REPORTED]: "Already Reported",
    [HttpStatus.IM_USED]: "IM Used",
    [HttpStatus.MULTIPLE_CHOICES]: "Multiple Choices",
    [HttpStatus.MOVED_PERMANENTLY]: "Moved Permanently",
    [HttpStatus.FOUND]: "Found",
    [HttpStatus.SEE_OTHER]: "See Other",
    [HttpStatus.NOT_MODIFIED]: "Not Modified",
    [HttpStatus.USE_PROXY]: "Use Proxy",
    [HttpStatus.TEMPORARY_REDIRECT]: "Temporary Redirect",
    [HttpStatus.PERMANENT_REDIRECT]: "Permanent Redirect",
    [HttpStatus.BAD_REQUEST]: "Bad Request",
    [HttpStatus.UNAUTHORIZED]: "Unauthorized",
    [HttpStatus.PAYMENT_REQUIRED]: "Payment Required",
    [HttpStatus.FORBIDDEN]: "Forbidden",
    [HttpStatus.NOT_FOUND]: "Not Found",
    [HttpStatus.METHOD_NOT_ALLOWED]: "Method Not Allowed",
    [HttpStatus.NOT_ACCEPTABLE]: "Not Acceptable",
    [HttpStatus.PROXY_AUTHENTICATION_REQUIRED]: "Proxy Authentication Required",
    [HttpStatus.REQUEST_TIMEOUT]: "Request Timeout",
    [HttpStatus.CONFLICT]: "Conflict",
    [HttpStatus.GONE]: "Gone",
    [HttpStatus.LENGTH_REQUIRED]: "Length Required",
    [HttpStatus.PRECONDITION_FAILED]: "Precondition Failed",
    [HttpStatus.PAYLOAD_TOO_LARGE]: "Payload Too Large",
    [HttpStatus.URI_TOO_LONG]: "URI Too Long",
    [HttpStatus.UNSUPPORTED_MEDIA_TYPE]: "Unsupported Media Type",
    [HttpStatus.RANGE_NOT_SATISFIABLE]: "Range Not Satisfiable",
    [HttpStatus.EXPECTATION_FAILED]: "Expectation Failed",
    [HttpStatus.MISDIRECTED_REQUEST]: "Misdirected Request",
    [HttpStatus.UNPROCESSABLE_ENTITY]: "Unprocessable Entity",
    [HttpStatus.LOCKED]: "Locked",
    [HttpStatus.FAILED_DEPENDENCY]: "Failed Dependency",
    [HttpStatus.TOO_EARLY]: "Too Early",
    [HttpStatus.UPGRADE_REQUIRED]: "Upgrade Required",
    [HttpStatus.PRECONDITION_REQUIRED]: "Precondition Required",
    [HttpStatus.TOO_MANY_REQUESTS]: "Too Many Requests",
    [HttpStatus.REQUEST_HEADER_FIELDS_TOO_LARGE]: "Request Header Fields Too Large",
    [HttpStatus.UNAVAILABLE_FOR_LEGAL_REASONS]: "Unavailable For Legal Reasons",
    [HttpStatus.INTERNAL_SERVER_ERROR]: "Internal Server Error",
    [HttpStatus.NOT_IMPLEMENTED]: "Not Implemented",
    [HttpStatus.BAD_GATEWAY]: "Bad Gateway",
    [HttpStatus.SERVICE_UNAVAILABLE]: "Service Unavailable",
    [HttpStatus.GATEWAY_TIMEOUT]: "Gateway Timeout",
    [HttpStatus.HTTP_VERSION_NOT_SUPPORTED]: "HTTP Version Not Supported",
    [HttpStatus.VARIANT_ALSO_NEGOTIATES]: "Variant Also Negotiates",
    [HttpStatus.INSUFFICIENT_STORAGE]: "Insufficient Storage",
    [HttpStatus.LOOP_DETECTED]: "Loop Detected",
    [HttpStatus.NOT_EXTENDED]: "Not Extended",
    [HttpStatus.NETWORK_AUTHENTICATION_REQUIRED]: "Network Authentication Required",
};

export function httpStatusTextByCode(httpStatusCode: number | string): string {
    if (!httpStatusCode || !httpStatusCode.toString) {
        // Not a number or a string
        throw new Error("Invalid httpStatusCode");
    }

    for (const key of Object.keys(httpStatusText)) {
        if (key === httpStatusCode.toString()) {
            return httpStatusText[key];
        }
    }

    throw new Error("Invalid httpStatusCode");
}


src/index.ts
/*
  HTTP codes
  See HTTP Status Code Registry:
  https://www.iana.org/assignments/http-status-codes/http-status-codes.xhtml
  
  And RFC 2324 - https://tools.ietf.org/html/rfc2324
*/

export enum status {
  HTTP_100_CONTINUE = 100,
  HTTP_101_SWITCHING_PROTOCOLS = 101,
  HTTP_102_PROCESSING = 102,
  HTTP_103_EARLY_HINTS = 103,
  HTTP_200_OK = 200,
  HTTP_201_CREATED = 201,
  HTTP_202_ACCEPTED = 202,
  HTTP_203_NON_AUTHORITATIVE_INFORMATION = 203,
  HTTP_204_NO_CONTENT = 204,
  HTTP_205_RESET_CONTENT = 205,
  HTTP_206_PARTIAL_CONTENT = 206,
  HTTP_207_MULTI_STATUS = 207,
  HTTP_208_ALREADY_REPORTED = 208,
  HTTP_226_IM_USED = 226,
  HTTP_300_MULTIPLE_CHOICES = 300,
  HTTP_301_MOVED_PERMANENTLY = 301,
  HTTP_302_FOUND = 302,
  HTTP_303_SEE_OTHER = 303,
  HTTP_304_NOT_MODIFIED = 304,
  HTTP_305_USE_PROXY = 305,
  HTTP_306_RESERVED = 306,
  HTTP_307_TEMPORARY_REDIRECT = 307,
  HTTP_308_PERMANENT_REDIRECT = 308,
  HTTP_400_BAD_REQUEST = 400,
  HTTP_401_UNAUTHORIZED = 401,
  HTTP_402_PAYMENT_REQUIRED = 402,
  HTTP_403_FORBIDDEN = 403,
  HTTP_404_NOT_FOUND = 404,
  HTTP_405_METHOD_NOT_ALLOWED = 405,
  HTTP_406_NOT_ACCEPTABLE = 406,
  HTTP_407_PROXY_AUTHENTICATION_REQUIRED = 407,
  HTTP_408_REQUEST_TIMEOUT = 408,
  HTTP_409_CONFLICT = 409,
  HTTP_410_GONE = 410,
  HTTP_411_LENGTH_REQUIRED = 411,
  HTTP_412_PRECONDITION_FAILED = 412,
  HTTP_413_REQUEST_ENTITY_TOO_LARGE = 413,
  HTTP_414_REQUEST_URI_TOO_LONG = 414,
  HTTP_415_UNSUPPORTED_MEDIA_TYPE = 415,
  HTTP_416_REQUESTED_RANGE_NOT_SATISFIABLE = 416,
  HTTP_417_EXPECTATION_FAILED = 417,
  HTTP_418_IM_A_TEAPOT = 418,
  HTTP_421_MISDIRECTED_REQUEST = 421,
  HTTP_422_UNPROCESSABLE_ENTITY = 422,
  HTTP_423_LOCKED = 423,
  HTTP_424_FAILED_DEPENDENCY = 424,
  HTTP_425_TOO_EARLY = 425,
  HTTP_426_UPGRADE_REQUIRED = 426,
  HTTP_428_PRECONDITION_REQUIRED = 428,
  HTTP_429_TOO_MANY_REQUESTS = 429,
  HTTP_431_REQUEST_HEADER_FIELDS_TOO_LARGE = 431,
  HTTP_451_UNAVAILABLE_FOR_LEGAL_REASONS = 451,
  HTTP_500_INTERNAL_SERVER_ERROR = 500,
  HTTP_501_NOT_IMPLEMENTED = 501,
  HTTP_502_BAD_GATEWAY = 502,
  HTTP_503_SERVICE_UNAVAILABLE = 503,
  HTTP_504_GATEWAY_TIMEOUT = 504,
  HTTP_505_HTTP_VERSION_NOT_SUPPORTED = 505,
  HTTP_506_VARIANT_ALSO_NEGOTIATES = 506,
  HTTP_507_INSUFFICIENT_STORAGE = 507,
  HTTP_508_LOOP_DETECTED = 508,
  HTTP_510_NOT_EXTENDED = 510,
  HTTP_511_NETWORK_AUTHENTICATION_REQUIRED = 511,
  /*
      WebSocket codes
      https://www.iana.org/assignments/websocket/websocket.xml#close-code-number
      https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent
    */
  WS_1000_NORMAL_CLOSURE = 1000,
  WS_1001_GOING_AWAY = 1001,
  WS_1002_PROTOCOL_ERROR = 1002,
  WS_1003_UNSUPPORTED_DATA = 1003,
  WS_1005_NO_STATUS_RCVD = 1005,
  WS_1006_ABNORMAL_CLOSURE = 1006,
  WS_1007_INVALID_FRAME_PAYLOAD_DATA = 1007,
  WS_1008_POLICY_VIOLATION = 1008,
  WS_1009_MESSAGE_TOO_BIG = 1009,
  WS_1010_MANDATORY_EXT = 1010,
  WS_1011_INTERNAL_ERROR = 1011,
  WS_1012_SERVICE_RESTART = 1012,
  WS_1013_TRY_AGAIN_LATER = 1013,
  WS_1014_BAD_GATEWAY = 1014,
  WS_1015_TLS_HANDSHAKE = 1015,
}

export default status;

export enum description {
  /*
    https://www.iana.org/assignments/http-status-codes/http-status-codes.xhtml#http-status-codes-1
  */
  HTTP_100_CONTINUE = "Continue",
  HTTP_101_SWITCHING_PROTOCOLS = "Switching Protocols",
  HTTP_102_PROCESSING = "Processing",
  HTTP_103_EARLY_HINTS = "Early Hints",
  HTTP_200_OK = "OK",
  HTTP_201_CREATED = "Created",
  HTTP_202_ACCEPTED = "Accepted",
  HTTP_203_NON_AUTHORITATIVE_INFORMATION = "Non-Authoritative Information",
  HTTP_204_NO_CONTENT = "No Content",
  HTTP_205_RESET_CONTENT = "Reset Content",
  HTTP_206_PARTIAL_CONTENT = "Partial Content",
  HTTP_207_MULTI_STATUS = "Multi-Status",
  HTTP_208_ALREADY_REPORTED = "Already Reported",
  HTTP_226_IM_USED = "IM Used",
  HTTP_300_MULTIPLE_CHOICES = "Multiple Choices",
  HTTP_301_MOVED_PERMANENTLY = "Moved Permanently",
  HTTP_302_FOUND = "Found",
  HTTP_303_SEE_OTHER = "See Other",
  HTTP_304_NOT_MODIFIED = "Not Modified",
  HTTP_305_USE_PROXY = "Use Proxy",
  HTTP_306_RESERVED = "(Unused)",
  HTTP_307_TEMPORARY_REDIRECT = "Temporary Redirect",
  HTTP_308_PERMANENT_REDIRECT = "Permanent Redirect",
  HTTP_400_BAD_REQUEST = "Bad Request",
  HTTP_401_UNAUTHORIZED = "Unauthorized",
  HTTP_402_PAYMENT_REQUIRED = "Payment Required",
  HTTP_403_FORBIDDEN = "Forbidden",
  HTTP_404_NOT_FOUND = "Not Found",
  HTTP_405_METHOD_NOT_ALLOWED = "Method Not Allowed",
  HTTP_406_NOT_ACCEPTABLE = "Not Acceptable",
  HTTP_407_PROXY_AUTHENTICATION_REQUIRED = "Proxy Authentication Required",
  HTTP_408_REQUEST_TIMEOUT = "Request Timeout",
  HTTP_409_CONFLICT = "Conflict",
  HTTP_410_GONE = "Gone",
  HTTP_411_LENGTH_REQUIRED = "Length Required",
  HTTP_412_PRECONDITION_FAILED = "Precondition Failed",
  HTTP_413_REQUEST_ENTITY_TOO_LARGE = "Content Too Large",
  HTTP_414_REQUEST_URI_TOO_LONG = "URI Too Long",
  HTTP_415_UNSUPPORTED_MEDIA_TYPE = "Unsupported Media Type",
  HTTP_416_REQUESTED_RANGE_NOT_SATISFIABLE = "Range Not Satisfiable",
  HTTP_417_EXPECTATION_FAILED = "Expectation Failed",
  HTTP_418_IM_A_TEAPOT = "(Unused)",
  HTTP_421_MISDIRECTED_REQUEST = "Misdirected Request",
  HTTP_422_UNPROCESSABLE_ENTITY = "Unprocessable Content",
  HTTP_423_LOCKED = "Locked",
  HTTP_424_FAILED_DEPENDENCY = "Failed Dependency",
  HTTP_425_TOO_EARLY = "Too Early",
  HTTP_426_UPGRADE_REQUIRED = "Upgrade Required",
  HTTP_428_PRECONDITION_REQUIRED = "Precondition Required",
  HTTP_429_TOO_MANY_REQUESTS = "Too Many Requests",
  HTTP_431_REQUEST_HEADER_FIELDS_TOO_LARGE = "Request Header Fields Too Large",
  HTTP_451_UNAVAILABLE_FOR_LEGAL_REASONS = "Unavailable For Legal Reasons",
  HTTP_500_INTERNAL_SERVER_ERROR = "Internal Server Error",
  HTTP_501_NOT_IMPLEMENTED = "Not Implemented",
  HTTP_502_BAD_GATEWAY = "Bad Gateway",
  HTTP_503_SERVICE_UNAVAILABLE = "Service Unavailable",
  HTTP_504_GATEWAY_TIMEOUT = "Gateway Timeout",
  HTTP_505_HTTP_VERSION_NOT_SUPPORTED = "HTTP Version Not Supported",
  HTTP_506_VARIANT_ALSO_NEGOTIATES = "Variant Also Negotiates",
  HTTP_507_INSUFFICIENT_STORAGE = "Insufficient Storage",
  HTTP_508_LOOP_DETECTED = "Loop Detected",
  HTTP_510_NOT_EXTENDED = "Not Extended (OBSOLETED)",
  HTTP_511_NETWORK_AUTHENTICATION_REQUIRED = "Network Authentication Required",
  /*
    https://www.iana.org/assignments/websocket/websocket.xml#close-code-number
  */
  WS_1000_NORMAL_CLOSURE = "Normal Closure",
  WS_1001_GOING_AWAY = "Going Away",
  WS_1002_PROTOCOL_ERROR = "Protocol error",
  WS_1003_UNSUPPORTED_DATA = "Unsupported Data",
  WS_1005_NO_STATUS_RCVD = "No Status Rcvd",
  WS_1006_ABNORMAL_CLOSURE = "Abnormal Closure",
  WS_1007_INVALID_FRAME_PAYLOAD_DATA = "Invalid frame payload data",
  WS_1008_POLICY_VIOLATION = "Policy Violation",
  WS_1009_MESSAGE_TOO_BIG = "Message Too Big",
  WS_1010_MANDATORY_EXT = "Mandatory Ext.",
  WS_1011_INTERNAL_ERROR = "Internal Error",
  WS_1012_SERVICE_RESTART = "Service Restart",
  WS_1013_TRY_AGAIN_LATER = "Try Again Later",
  WS_1014_BAD_GATEWAY = "The server was acting as a gateway or proxy and received an invalid response from the upstream server.",
  WS_1015_TLS_HANDSHAKE = "TLS handshake",
}
